shared/actions/devices.js

```diff
@@ -1,138 +1,192 @@
+      error: true,
+    }: ShowDevices))
   }
 }
 
-export function removeDevice (deviceID: string, name: string, currentDevice: boolean): AsyncAction {
-  return (dispatch, getState) => {
-    if (currentDevice) {
-      // Revoking the current device uses the "deprovision" RPC instead.
-      const username = getState().config.username
+function * _deviceRemoveSaga (removeAction: RemoveDevice): SagaGenerator<any, any> {
+  // Revoking the current device uses the "deprovision" RPC instead.
+  const {currentDevice, name, deviceID} = removeAction.payload
+  if (currentDevice) {
+    try {
+      const username = yield select(state => state.config && state.config.username)
       if (!username) {
-        console.warn('No username in removeDevice')
-        return
+        const error = {errorText: 'No username in removeDevice'}
+        console.warn(error)
+        yield put(({
+          type: Constants.deviceRemoved,
+          payload: error,
+          error: true,
+        }: DeviceRemoved))
       }
-      loginDeprovisionRpc({
-        param: {username, doRevoke: true},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
-
-          if (!error) {
-            dispatch(loadDevices())
-            dispatch(setRevokedSelf(name))
-            dispatch(navigateTo([], loginTab))
-            dispatch(switchTab(loginTab))
-          }
-        },
-      })
-    } else {
-      revokeRevokeDeviceRpc({
+      yield call(loginDeprovisionRpcPromise, {param: {username, doRevoke: true}})
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
+      }: DeviceRemoved))
+      yield put(loadDevices())
+      yield put(setRevokedSelf(name))
+      yield put(navigateTo([], loginTab))
+      yield put(switchTab(loginTab))
+    } catch (e) {
+      console.warn('Error removing the current device:', e)
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: e,
+        error: true,
+      }: DeviceRemoved))
+    }
+  } else {
+    // Not the current device.
+    try {
+      yield call(revokeRevokeDeviceRpcPromise, {
         param: {deviceID, force: false},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
+      })
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
```

Comment:
Oh, cool.  I actually kind of like explicitly marking it as a success path that way now that I'm considering removing it!